// total/dist/lib/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class SubcommandNotFound extends Error {
    constructor(attempt, ...args) {
        super(`error: unknown subcommand \`${attempt}\``);
        this.attempt = attempt;
        this.name = 'SubcommandNotFound';
    }
}
exports.SubcommandNotFound = SubcommandNotFound;
class FileNotFound extends Error {
    constructor(attempt, ...args) {
        super(`error: file \`${attempt}\` not found`);
        this.attempt = attempt;
        this.name = 'FileNotFound';
    }
}
exports.FileNotFound = FileNotFound;
class PathNotFound extends Error {
    constructor(attempt, ...args) {
        super(`error: path \`${attempt}\` not found`);
        this.attempt = attempt;
        this.name = 'PathNotFound';
    }
}
exports.PathNotFound = PathNotFound;
class InvalidArguments extends Error {
    constructor(attempt = '', ...args) {
        super(`error: invalid arguments`);
        this.attempt = attempt;
        this.name = 'InvalidArguments';
    }
}
exports.InvalidArguments = InvalidArguments;
